File: src/service/session_service.go
/*
 * Auth: session_service.go
 * Copyright (C) 2025 mtctx
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the **GNU General Public License** as published
 * by the Free Software Foundation, either **version 3** of the License, or
 * (at your option) any later version.
 *
 * *This program is distributed WITHOUT ANY WARRANTY;** see the
 * GNU General Public License for more details, which you should have
 * received with this program.
 *
 * SPDX-FileCopyrightText: 2025 mtctx
 * SPDX-License-Identifier: GPL-3.0-only
 */

package service

import (
	"errors"
	"fmt"
	"src/crypto"
	"src/data"
	"strings"
	"time"

	"github.com/google/uuid"
	"github.com/rs/zerolog/log"
)

func GetSession(setup GeneralData[data.Session], uuid uuid.UUID) (data.Session, error) {
	err := setup.Validate()
	if err != nil {
		return data.Session{}, err
	}

	if value, found := setup.Cache.Get(uuid[:]); found {
		return value, nil
	}

	const query = "SELECT * FROM sessions WHERE uuid = $1"

	var session data.Session
	err = setup.Database.QueryRow(setup.DBContext, query, uuid).Scan(&session)
	if err != nil {
		return data.Session{}, err
	}

	if session.Expired() {
		return data.Session{}, errors.New("session expired")
	}

	return session, nil
}

func GetSessionMACKey(setup GeneralData[data.Session], uuid uuid.UUID) ([crypto.MacKeySize]byte, error) {
	err := setup.Validate()
	if err != nil {
		return [crypto.MacKeySize]byte{}, err
	}

	query := "SELECT mac_key FROM sessions WHERE uuid = $1"
	var macKey [crypto.MacKeySize]byte
	err = setup.Database.QueryRow(setup.DBContext, query, uuid).Scan(&macKey)
	if err != nil {
		return [crypto.MacKeySize]byte{}, err
	}

	return macKey, nil
}

func CreateSession(setup GeneralData[data.Session], ownerUUID uuid.UUID, validFor time.Duration) (data.Session, []byte, error) {
	err := setup.Validate()
	if err != nil {
		return data.Session{}, nil, err
	}

	key, err := crypto.NewMACKey()
	if err != nil {
		return data.Session{}, nil, err
	}
	generatedUUID := uuid.New()
	mac := crypto.ComputeMAC([]byte(fmt.Sprintf("%s:%d", generatedUUID, validFor)), key)

	session := data.Session{
		UUID:      generatedUUID,
		MacKey:    [crypto.MacKeySize]byte(key),
		CreatedAt: time.Now().UnixNano(),
		ValidFor:  validFor,
	}

	query := "INSERT INTO sessions (uuid, user_uuid, mac_key, created_at, valid_for) VALUES ($1, $2, $3, $4, $5)"
	_, err = setup.Database.Exec(setup.DBContext, query, session.UUID, ownerUUID, session.MacKey, session.CreatedAt, session.ValidFor)
	if err != nil {
		return data.Session{}, nil, err
	}

	added := setup.Cache.Set(generatedUUID[:], session, data.SessionCost)
	if !added {
		log.Error().Msg("Failed to add session to cache")
	}

	return session, mac, nil
}

func DeleteSession(setup GeneralData[data.Session], uuid uuid.UUID) error {
	err := setup.Validate()
	if err != nil {
		return err
	}

	setup.Cache.Del(uuid[:])

	query := "DELETE FROM sessions WHERE uuid = $1"
	_, err = setup.Database.Exec(setup.DBContext, query, uuid)
	return err
}

func SplitUUIDAndMAC(uuidAndMac string) (uuid.UUID, []byte, error) {
	result := strings.Split(uuidAndMac, ".")

	if len(result) != 2 {
		return uuid.Nil, []byte{}, errors.New("invalid uuid and mac")
	}

	parsedUUID, err := uuid.Parse(strings.TrimPrefix(result[0], "VA:"))
	if err != nil {
		return uuid.Nil, []byte{}, err
	}

	macTag := []byte(result[1])
	return parsedUUID, macTag, nil
}

func BuildSessionString(uuid uuid.UUID, macTag []byte) string {
	return fmt.Sprintf("VA:%s.%s", uuid, macTag)
}

File: src/main/kotlin/dev/vayen/service/SessionService.kt
/*
 * Vayen Auth (Vayen_Auth.main): SessionService.kt
 * Copyright (C) 2025 mtctx
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the **GNU General Public License** as published
 * by the Free Software Foundation, either **version 3** of the License, or
 * (at your option) any later version.
 *
 * This program is distributed WITHOUT ANY WARRANTY; see the
 * GNU General Public License for more details, which you should have
 * received with this program.
 *
 * SPDX-FileCopyrightText: 2025 mtctx
 * SPDX-License-Identifier: GPL-3.0-only
 */

package dev.vayen.service

import com.mayakapps.kache.ObjectKache
import dev.vayen.config.Config
import dev.vayen.data.Session
import dev.vayen.table.SessionTable
import mtctx.lumina.v4.Lumina
import mtctx.utilities.*
import mtctx.utilities.crypto.HmacSha256
import org.jetbrains.exposed.sql.Database
import org.jetbrains.exposed.sql.SqlExpressionBuilder.eq
import org.jetbrains.exposed.sql.deleteWhere
import org.jetbrains.exposed.sql.insert
import org.jetbrains.exposed.sql.selectAll
import java.util.*
import kotlin.time.Clock
import kotlin.time.Duration
import kotlin.time.ExperimentalTime
import kotlin.uuid.ExperimentalUuidApi

@OptIn(ExperimentalTime::class)
class SessionService(
    config: Config, lumina: Lumina,
    database: Database, cache: ObjectKache<UUID, Session>
) : DatabaseService<UUID, Session>(config, lumina, database, cache) {
    suspend fun getAndCheck(uuid: UUID, macTag: ByteArray): Outcome<Session> = get(uuid).flatMap { session ->
        return if (HmacSha256.verify(toMacTagFormat(uuid, session.validFor), macTag, session.macKey)) success(session)
        else failure("Invalid MAC Tag!")
    }

    override suspend operator fun get(id: UUID): Outcome<Session> = runCatchingOutcomeOf<Session> {
        val session = cache.get(id)
            ?: dbQuery {
                SessionTable.selectAll().where { SessionTable.uuid eq id }.map {
                    val macTag = generateMacTag(id, it[SessionTable.validFor], it[SessionTable.macKey])

                    Session(
                        uuid = it[SessionTable.uuid],
                        userUUID = it[SessionTable.userUUID],
                        macKey = it[SessionTable.macKey],
                        createdAt = it[SessionTable.createdAt],
                        validFor = it[SessionTable.validFor],
                        cookieValue = buildSessionString(id, macTag)
                    )
                }.singleOrNull()
            } ?: return failure("Session not found")

        if (session.expired()) {
            return failure("Session Expired!")
        }
        session
    }

    suspend fun macKey(uuid: UUID): Outcome<ByteArray> = runCatchingOutcomeOf<ByteArray> {
        dbQuery {
            SessionTable.select(SessionTable.macKey).where { SessionTable.uuid eq uuid }.singleOrNull()
        }?.get(SessionTable.macKey)
            ?: throw IllegalStateException("Could not get the Mac Key from Session with UUID $uuid")
    }


    @OptIn(ExperimentalUuidApi::class)
    suspend fun create(ownerUUID: UUID): Outcome<Session> = runCatchingOutcomeOf<Session> {
        val generatedMacKey = HmacSha256.generateKey()
        val generatedUUID = UUID.randomUUID()
        val now = Clock.System.now()

        dbQuery {
            SessionTable.insert {
                it[uuid] = generatedUUID
                it[userUUID] = ownerUUID
                it[macKey] = generatedMacKey
                it[createdAt] = now
                it[validFor] = config.session.validFor
            }
        }

        val sessionString =
            buildSessionString(generatedUUID, generateMacTag(generatedUUID, config.session.validFor, generatedMacKey))
        val createdSession =
            Session(generatedUUID, ownerUUID, generatedMacKey, now, config.session.validFor, sessionString)
        cache.put(generatedUUID, createdSession)
        return success(createdSession)
    }

    override suspend fun delete(id: UUID) = runCatchingOutcomeOf<Unit> {
        cache.remove(id)
        dbQuery {
            SessionTable.deleteWhere { SessionTable.uuid eq id }
        }
    }

    companion object {
        suspend fun splitUUIDAndMAC(uuidAndMac: String): Outcome<Pair<UUID, ByteArray>> =
            runCatchingOutcomeOf<Pair<UUID, ByteArray>> {
                val result = uuidAndMac.split(".")

                if (result.size != 2) {
                    return failure("invalid uuid and mac")
                }

                val parsedUUID = UUID.fromString(result[0].removePrefix("VA:"))
                val macTag = result[1].chunked(2).map { it.toInt(16).toByte() }.toByteArray()
                return success(parsedUUID to macTag)
            }

        fun buildSessionString(uuid: UUID, macTag: ByteArray): String =
            "VA:$uuid.${macTag.joinToString("") { "%02x".format(it) }}"

        fun toMacTagFormat(uuid: UUID, validFor: Duration): String = "$uuid:$validFor"
        fun generateMacTag(uuid: UUID, validFor: Duration, macKey: ByteArray): ByteArray =
            HmacSha256.generate(toMacTagFormat(uuid, validFor), macKey).tag
    }
}